Mk transifex integration

mkangia opened the pull request:
[Feature changes](https://docs.google.com/document/d/1_FtX-lsEfaNVo2wFQ8Ob6E5albXyV8n9iOddxkIXTbU/edit#bookmark=id.9w722d9qkl6m)

[Non-feature related changes](https://docs.google.com/document/d/1_FtX-lsEfaNVo2wFQ8Ob6E5albXyV8n9iOddxkIXTbU/edit#bookmark=id.7w2p29o5dzrq)

best review 🐟 / 🐟 

a quick overview of the two views(Pull resource and convert translations) added in [video](http://recordit.co/OZxLJ3l7WU)
the link `/a/<domain>/apps/translations` is not added anywhere on HQ as a link.

mkangia committed:
extract POFileGenerator

mkangia committed:
add view to convert xlsx to po and vice versa

mkangia committed:
upgrade translations as an app and plug in app translations from icds

mkangia committed:
add more languages

mkangia committed:
use slug instead of name

slug was considered mostly to be same as name but its not a safe assumption

mkangia committed:
add ability to pull a resource

mkangia committed:
remove redundancies

- remove reduntant argument
- dont make button lg/large
- dont disable button.  revisit this if need to disable it, currently it does not re-enable

mkangia committed:
avoid skipping if index is 0 which is evaluated as False

mkangia committed:
add comments and extract get_file_content_from_workbook

mkangia committed:
better naming and move clean up to PoFileGenerator to use it

mkangia committed:
lint + comment

mkangia committed:
revert changes already done in another PR

mkangia committed:
merge master

mkangia committed:
provide all possible languages

mkangia committed:
add other languages which differ on transifex from HQ [skip ci]

mkangia committed:
lint [skip ci]

mkangia committed:
lint [skip ci]

mkangia commented on corehq/apps/translations/forms.py:
```diff
@@ -0,0 +1,86 @@
+    transifex_project_slug = forms.ChoiceField(label=ugettext_lazy("Trasifex project"), choices=(),
+                                               required=True)
+    target_lang = forms.ChoiceField(label=ugettext_lazy("Target Language"),
+                                    choices=langcodes.get_all_langs_for_select(),
```
cache return https://github.com/dimagi/langcodes/pull/22

millerdev commented on corehq/apps/translations/views.py:
```diff
@@ -0,0 +1,195 @@
+        """
+        translations = self._generate_translations_for_po(worksheet)
+        po_file_generator = PoFileGenerator()
+        po_file_generator.generate_translation_files(translations, {})
```
Is there a reason not to generate translations as part of the constructor? This could be shortened to
```py
po_file_generator = PoFileGenerator(translations, {})
```
As is, `PoFileGenerator` is in an invalid state on initial construction until `generate_translation_files` is called. This is an OO anti-pattern. Some things that might invalidate this suggestion:

- if `generate_translation_files` is called more than once for a given `PoFileGenerator` instance.
- if a `PoFileGenerator` instance is ever used without calling it's `generate_translation_files` method.

mkangia replied:
Does [this](https://github.com/dimagi/commcare-hq/blob/425d72efd19083a79e37c34643fc5d002528af63/corehq/apps/app_manager/app_translations/generators.py#L49) look like a valid invalidation? :)

millerdev replied:
Not necessarily. I can't tell if `self.po_file_generator` is referenced anywhere before `self.po_file_generator.generate_translation_files()` is called, and if it is, is its state (not having generated translation files yet) intentional at that point, or is it a bug to have referenced it before translations have been generated?

If it is not referenced before `generate_translation_files()` is called, then why does `self.po_file_generator` need to be created/assigned beforehand?

mkangia replied:
`po_file_generator` is used at [other places](https://github.com/dimagi/commcare-hq/blob/0a04cd5ffa39bc2a0e5825bdf49d46c3a5bbac6e/corehq/apps/app_manager/app_translations/generators.py#L191-L200) so just wanted to keep a reference on the object.
Also personally don't like `__init__` doing something major. I like `__init__` to just initialise things and be simple. But if its not with the OO pattern or not the python/django way of doing things, happy to modify it like you suggested.

millerdev replied:
There's a ton of conflicting information out there on this topic. Some people say the constructor should do the bare minimum, basically initialize fields. Others say it should do whatever work is necessary to get the object into a coherent state so other methods will return sensible values (not requiring another special method to be called to further initialize it before it's ready to be used). I think the design I'm hoping for is better thought of in terms of functional concepts.

With the design as is, the instance is not fully initialized when it is first constructed:
```py
po_file_generator = PoFileGenerator()
```
Notice it is possible, but makes no sense to reference `po_file_generator.generated_files` before calling
```py
po_file_generator.generate_translation_files(translations, {})
```
This does a second initialization. Before `generate_translation_files` is called the object is in a somewhat indeterminate state. Taking this further, how many times can or must I call `generate_translation_files`? Zero? One? What if I call it twice? If you move it into the constructor you don't have that problem because it's unconventional to call the constuctor any number of times other than once for a given instance.

Another way to handle this is to split your `PoFileGenerator` into two objects, so it can be used like this:
```py
po_file_generator = PoFileGenerator()
result = po_file_generator.generate_translation_files(translations, {})
try:
    for filename, filepath in result:
        # do stuff with filename/filepath
finally:
    result.cleanup()
```
In this case `PoFileGenerator` doesn't really need to be an object at all since it only has one method: it can just be a function (`generate_translation_files`) that returns the generated translation files. The `result` object is the more interesting part that will likely be a new object type (maybe call it `PoFiles`?) because it is both iterable and has a cleanup method. 

Does that clear things up?

esoergel replied:
I also think that classes should receive all the information they need on initialization, so they're never in a nonsensical state.  I _definitely_ also think that functions should be preferred over objects, as they are the simplest possible logical unit - you provide inputs and get an output.  Of course there are instances where classes make more sense.

To your point about avoiding work in `__init__`, that also makes a lot of sense.  Daniel has some good suggestions, but I'll throw in another to specifically address that.  You can provide all of the inputs to the `__init__` method, but perform any non-trivial computation lazily.
```python
class PoFileGenerator(object):
    def __init__(self, translations, metadata):
        self.translations = translations
        self.metadata = metadata

    @cached_property
    def generated_files(self):
        # body of generate_translation_files
```
Depending on which of the options you chose, the bulk of the work occurs:

1. During `PoFileGenerator` initialization, which isn't great
1. As currently implemented, during `po_file_generator.generate_translation_files`, which is certainly obvious.  This has the downside of of `po_file_generator` being in a partially initialized state and later mutated.
1. In my example above, when `po_file_generator.generated_files` is first accessed.  I think this is better than in the constructor, but still arguably surprising.  You could use a memoized `generate_translation_files()` method instead of a property if you want to be clear that work is happening.
1. If this were implemented as a function, then when that function is called.  To my mind, this is the best option, as you provide all args up front, and it's clear that you expect work to be done.  You then pass around the results, rather than an object which maintains the results.  The limitation is of course is that sometimes you just need a class.

mkangia replied:
@millerdev 
> Some people say the constructor should do the bare minimum

i think thats the group i belong to. i would still prefer keeping things simple.
tried [this](https://github.com/dimagi/commcare-hq/pull/21516/commits/acdf4f4812b58c9196f9ffe7c091af4cff145dcb) from the couple of suggestions you had. all looks okay but had to add [this](https://github.com/dimagi/commcare-hq/pull/21516/commits/acdf4f4812b58c9196f9ffe7c091af4cff145dcb#diff-389860842dade5566b99beb0b7021f2eR51) catch which is probably similar to what we are trying to resolve here.
Does this resolve the conflict?

millerdev commented on corehq/apps/translations/views.py:
```diff
@@ -0,0 +1,195 @@
+        worksheet = openpyxl.load_workbook(uploaded_file).worksheets[0]
+        po_file_generator = self._generate_po_file(worksheet)
+        content = open(po_file_generator.generated_files[0][1], 'r', encoding="utf-8").read()
+        po_file_generator.cleanup()
```
Seems like the cleanup should happen in a finally block. Also, proper file handling is a good idea. Two options:
```py
po_file_generator = self._generate_po_file(worksheet)
try:
    with open(po_file_generator.generated_files[0][1], 'r', encoding="utf-8") as fh:
        content = fh.read()
finally:
    po_file_generator.cleanup()
```
Or if you want to get more fancy you could make `po_file_generator` into a context manager and use it in a `with` block where `cleanup` gets called on `__exit__`.

mkangia replied:
Okay. I never thought read operation would fail unless in some very unexpected scenario which would not be very harmful. Will add the block just to be on a safer side.

millerdev commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -39,15 +39,16 @@ def __init__(self, domain, app_id, source_lang, project_slug, version=None, lang
-            self.po_file_generator.generate_translation_files()
+            self.transifex_po_file_generator.generate_translation_files()
             file_uploads = self._send_files_to_transifex()
             self._cleanup()
```
Looks like you're reinventing `try`/`finally` here. This code can be simplified:
```py
pogen = self.transifex_po_file_generator
try:
    pogen.generate_translation_files()
    file_uploads = self._send_files_to_transifex()
finally:
    pogen.po_file_generator.cleanup()
```
And then you can remove the `_cleanup` method.

Also it's probably a good idea to add a `cleanup` method on `TransifexPOFileGenerator` that calls `self.po_file_generator.cleanup()` so you don't have to reach into its internal state to do cleanup here.

millerdev commented:
Forgot to mention in my review, are there tests for `PoFileGenerator`? If not, would be a good idea to add them.

esoergel commented on corehq/apps/translations/urls.py:
```diff
@@ -0,0 +1,17 @@
+        name=ConvertTranslations.urlname),
+    url(r'^pull_resource/$', PullResource.as_view(),
+        name=PullResource.urlname),
+    url(r'^', ConvertTranslations.as_view(),
```
I think it should be `url(r'^$', ...)`
`url(r'^', ...)` will match anything that starts with `/a/<domain>/apps/translations`.  This pattern is typically used for defining a collection of URLs that all share a prefix - you'd use it with an `includes(...)` statement.

This is the last pattern in the file, so it'll just serve as a catchall as-is, but it's a potential source of future headaches.

mkangia replied:
> url(r'^', ...) will match anything that starts with /a/<domain>/apps/translations
thanks. will fix that. `url(r'^$', ...)` matches both `/a/<domain>/apps/translations` and `/a/<domain>/apps/translations/`

Just intended to have one of the always available view as the default view, so it has its own link and also added for the parent URL.
Like if this is added as a tab/under a tab the link would be `/a/<domain>/apps/translations` and would show the view for `/a/<domain>/apps/translations/convert_translations` as well.
Actually i can just remove the first link as well. Will do that.

esoergel commented on custom/icds_reports/views.py:
```diff
@@ -1599,6 +1600,7 @@ class ICDSAppTranslations(BaseDomainView):
     page_title = ugettext_lazy('ICDS App Translations')
     urlname = 'icds_app_translations'
     template_name = 'icds_reports/icds_app/app_translations.html'
+    section_name = ugettext_noop("Translations")
```
Shouldn't this be `ugettext_lazy` like `page_title`?

mkangia replied:
Should this always be `ugettext_lazy`. we have used `ugettext_noop` at places for section names as well.

esoergel replied:
Oh, actually, you're right - looks like `ugettext_noop` should work:
https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/reports/templates/reports/standard/base_template.html#L69-L69
It should only be used in variables which are explicitly translated later.

I'm not convinced we need to use `ugettext_noop` in this instance, and doing so may mean that it doesn't get translated when the report is rendered through other means, like when it's exported or emailed, unless it's also translated there (I didn't see any other relevant matches for `trans.*section_name`, `_\(.*section_name`, or `ugettext\(.*section_name`).  That however, is a battle for another day, as this is clearly meant to accept `ugettext_noop`.

mkangia committed:
extract BaseTranslationsView and show links for transifiex only when needed

mkangia commented on corehq/apps/translations/views.py:
```diff
@@ -0,0 +1,121 @@
+        uploaded_file = self.convert_translation_form.cleaned_data.get('upload_file')
+        worksheet = openpyxl.load_workbook(uploaded_file).worksheets[0]
+        po_file_generator = self._generate_po_file(worksheet)
+        content = open(po_file_generator.generated_files[0][1], 'r', encoding="utf-8").read()
```
i thought since its not stored to an object, there wont be any open references left?

esoergel replied:
I don't know the answer to that - I know that `with open()` is fine and standard practice, but I'm not certain about `open()`.  I think it's one of those things that depends on the architecture and the python interpreter being used, as opposed to `with`, which does it as part of the language specification.  That is, it probably does do what you want, but it still feels weird to me.  Not that big a deal though - I think it's a style thing that you can decide.

millerdev replied:
`with open()` is more explicit in terms of resource cleanup, and therefore is better. Using `open()` without explicitly closing the file means that the file reference hangs around as long as there is a name referencing it in some scope or garbage collection hasn't been run on the names that were referencing it. On CPython, garbage collection is somewhat deterministic because of the way reference counting is implemented, so in practice using bare `open()` will most likely not hurt anything—although it still could if the code was written in a way that hit max open file count or something like that. Most other versions of Python do not have the same guarantees with respect to garbage collection.

It is a best practice to always do explicit resource cleanup (use `with open()`) because that makes it easier to reason about the resources your code is using.

mkangia replied:
Noted. thank you 👍

esoergel commented on corehq/apps/translations/views.py:
```diff
@@ -21,10 +23,27 @@
+class BaseTranslationsView(BaseDomainView):
+    @property
+    def page_context(self):
+        transifex_details_available = transifex_details_available_for_domain(self.domain)
```
This might be nice as a memoized property, since it's referenced in a few other places

esoergel commented on corehq/apps/translations/views.py:
```diff
@@ -132,10 +151,13 @@ def section_url(self):
 
 
-class PullResource(BaseDomainView):
+@method_decorator([toggles.APP_TRANSLATIONS_WITH_TRANSIFEX.required_decorator()], name='dispatch')
```
There are 3 booleans used in these views: `toggles.APP_TRANSLATIONS_WITH_TRANSIFEX`, `transifex_details_available`, and `transifex_integration_enabled`.  It looks like at least the last two are equivalent - it'd be simpler to understand if you either consolidated those or used the same term to indicate that they mean the same thing.

mkangia replied:
> toggles.APP_TRANSLATIONS_WITH_TRANSIFEX
ensures user permission to access the page.

> transifex_details_available
ensures there are settings available for the project and is available as a context variable

> transifex_integration_enabled
is just a view level method which confirms transifex settings and adds an error message

probably not an improvement, but an attempt [here](https://github.com/dimagi/commcare-hq/pull/21516/commits/b0a4ea2049bf2d59fb9f4fd718f3204feeddf6f0)

esoergel replied:
That change is helpful.  I was also thinking that `self.transifex_integration_enabled` is the same as `self.transifex_details_available`, except that the former has a side effect of adding an error message to the request.  I found that confusing, since the names sound different.

One alternative would be to drop the side effects and move the error to the places where `transifex_integration_enabled` is used.  This way it would be clear that the same check is used consistently, and when reading the `post` method, it's clear that an error message is displayed when the details are not available.

mkangia committed:
remove duplicate url and restrict root url for translations app

mkangia committed:
refine logic to fetch resource slugs

mkangia committed:
remove mappings for garo, khasi and mizo and add for sindhi

mkangia committed:
add ability to move resources

Used this to transfer resources from one project to another. Not exposed via HQ but can come in handy for dev purposes.

mkangia committed:
extract logic to a single place on the view

mkangia committed:
memoize

mkangia committed:
use with open instead

mkangia committed:
keep it simple, index unused

orangejenny commented on corehq/apps/translations/templates/convert_translations.html:
```diff
@@ -0,0 +1,15 @@
+{% load hq_shared_tags %}
+{% load i18n %}
+{% load compress %}
+{% load djangular_tags %}
```
This isn't used...actually, most of these aren't used, the angular one just caught my eye because we're deprecating angular.

mkangia replied:
Thanks, was meant to remove this earlier.

orangejenny commented on corehq/apps/translations/templates/translations/base_template.html:
```diff
@@ -0,0 +1,20 @@
+                {% trans "Transifex Integration" %}
+            </a>
+        </li>
+    </ul>
```
Yes, please - looking at `ConvertTranslations` at least it looks like you already have `page_title` and `section_name` set, so it might just work or you might need to add in `parent_properties` or something like that.

mkangia replied:
@orangejenny @esoergel not sure whats the change needed here. 
Also i have not added this as/under a tab anywhere yet in html since all of this is still for custom. Also was not sure where to add this in dropdowns available but under app seems like the best places for now and hence the urls are under `/a/<domain>/apps/`. So for now thought to keeping it a secret url that is not available as a link on HQ.

orangejenny replied:
Sorry, I misread Ethan's comment. He & I are talking about two different things, but they're both about making the UI code more standard.

Even though this is custom, it's still better to add to a tab, so long as it's hidden behind the appropriate toggle/settings, rather than having bespoke UI code and secret URLs (people are bad at remembering them). I think that you can add a new tab with the `apps/translations` prefix and as long as you don't add it to [MENU_TABS](https://github.com/dimagi/commcare-hq/blob/master/corehq/tabs/config.py) it won't show in the top nav. Then you could add a general "Translations" link in the apps menu, probably under "New Application". You'll also need this template to extend `two_column.html` so the sidebar is created for you.

I was talking about the breadcrumbs, which get set based on `section_name`, `parent_pages`, etc.

esoergel replied:
Yeha, I was talking about the sidebar which is typically declared in that `tabclasses.py` file and automatically generated in `two_column.html`.  That basically forms a site map of all of HQ.

Here's a recent example where I made a report appear on the domain admin page when the corresponding feature flag was enabled:
https://github.com/dimagi/commcare-hq/blob/075bfbb5bfe358eb4fffaef4c8df2330e0300009/corehq/tabs/tabclasses.py#L1585-L1590

Note also that the dropdown menu is also controlled by this interface - you can add these things there by adding `'show_in_dropdown': True`

mkangia replied:
Thanks both,
[here](https://github.com/dimagi/commcare-hq/pull/21516/commits/f7b994e61c68b6f046f9569714fd488461fd412e) is my attempt for this.
I have added this as a separate link under Applications Tab if the feature flag for transifex integration is enabled which is at a user level.

Couple of things:
1. Added the link under "Applications" tab for translations.
2. The applications tab is not highlighted when visiting these pages which i think should be expected. I think the only way around is to add another tab as "Translations" and handle its view permission (but "Translations" itself does not look like something that should be a tab) or to edit the ApplicationsTab itself itself instead of adding a new tab. Should that be needed or its okay for now?

orangejenny replied:
That looks good to me.

> The applications tab is not highlighted

I thought you could do this by making `url_prefix_formats` something like `/a/{domain}/apps/translations/` (and updating urls.py to correspond, of course), but if you've already futzed with this and it isn't working, I think it's fine that the tab isn't highlighted.

mkangia replied:
I tried that. If i do that the `ApplicationTab` becomes the active tab and the `TranslationsTab` does not get picked and hence the side bar is not added.

mkangia committed:
use ugettext_lazy

mkangia committed:
remove unused tags

orangejenny commented on corehq/apps/translations/static/translations/js/setup.js:
```diff
@@ -0,0 +1,11 @@
+hqDefine('translations/js/setup', [
```
Hooray for files being requirejs-compatible right out of the gate.

FWIW, since this file is only doing select2 initialization, instead of creating this file you could add the class `ko-select2` to the two inputs and then use `hqwebapp/js/widgets` as `requirejs_main`.

mkangia replied:
my first requirejs-compatible js contribution removed thought.. 
some day..

orangejenny replied:
There are a lot of migration tasks if you would like one ;)

mkangia committed:
add clean up for generator and use that instead, use try finally

mkangia committed:
ensure files are cleaned up

orangejenny commented on corehq/apps/translations/templates/translations/base_template.html:
```diff
@@ -12,18 +12,22 @@ <h2 class="text-hq-nav-header">{% trans "TOOLS" %}</h2>
-        </li>
-    </ul>
+    {% if transifex_details_available %}
+        {% if request|toggle_enabled:"APP_TRANSLATIONS_WITH_TRANSIFEX" %}
```
It's more typical to use the toggle's `required_decorator` on the view, and then you also get a handy [message](https://github.com/dimagi/commcare-hq/pull/15485/) if the toggle isn't enabled.

mkangia replied:
These is the base template which has links(one link for now, convert translations) which does not need this flag enabled. So its just to hide links that need the flag enabled. 
The views corresponding to the links that need the flag have the check at the view level.

orangejenny reviewed (commented):
I like the approach of having a separate translations app. Looks good to me once Ethan and Daniel's comments are addressed.

mkangia committed:
just use class and available widget instead of adding a new js

mkangia committed:
remove redundant tags

mkangia committed:
add tests for POFileGenerator

mkangia commented:
> Forgot to mention in my review, are there tests for PoFileGenerator? If not, would be a good idea to add them.

@millerdev added.

mkangia commented:
@esoergel @millerdev request for a review. Tried addressing all concerns.

mkangia committed:
lint

millerdev commented on custom/icds/translations/integrations/client.py:
```diff
@@ -196,3 +196,39 @@ def get_source_lang(self):
+            }
+            upload_response = requests.put(
+                upload_url, data=json.dumps(data), auth=self._auth, headers=headers,
+            )
```
ditto on error handling

millerdev reviewed (commented):
Here's an example of how `POFileGenerator` could be structured to make its state more externally consistent:
```py
class POFileGenerator(object):

    def __init__(self, translations, metadata):
        self.generated_files = []  # list of tuples (filename, filepath)
        try:
            self._generate_translation_files(translations, metadata)
        except Exception:
            exc_info = sys.exc_info()
            self.cleanup()
            six.reraise(*exc_info)

    def _generate_translation_files(self, translations, metadata):
        # note this is now a private method
        ...

    def cleanup(self):
        ...  # remove generated files here

        # clear obsolete references
        self.generated_files = []

    def __enter__(self):
        return self

    def __exit__(self, *exc_info):
        self.cleanup()
```

Then you can use it like this:

```py
with POFileGenerator(trans, meta) as po_file_generator:
    ...  # do stuff with generated translations
```

Or like this

```py
po_file_generator = POFileGenerator(trans, meta)
try:
    ...  # do stuff with generated translations
finally:
    po_file_generator.cleanup()
```

esoergel commented on custom/icds/translations/integrations/client.py:
```diff
@@ -40,10 +40,17 @@ def get_resource_slugs(self, version):
+        elif version and not self.use_version_postfix:
+            # get all slugs that don't have version postfix
+            return [r['slug']
+                    for r in self.list_resources().json()
```
nit: you can use `all_resources` here and on line 45 rather than `self.list_resources().json()`

esoergel reviewed (approved):
There's some pending feedback, but I don't think I had any remaining blockers

mkangia commented on custom/icds/translations/integrations/client.py:
```diff
@@ -196,3 +196,39 @@ def get_source_lang(self):
+            url = "https://www.transifex.com/api/2/project/{}/resource/{}/translation/{}/?file".format(
+                self.project, resource_slug, lang
+            )
+            response = requests.get(url, auth=self._auth, stream=True)
```
This method is currently not called anywhere but was used for some dev work and was included because its useful.
I think its fine if it raises an exception.

mkangia replied:
also the other places where we request to transifex is either as a celery task or within a POST Http request to HQ from the browser for a quick check, so both should be fine, considering celery task would just retry/fail eventually and the request to HQ would just throw an exception but would probably succeed on retry.
Also, i don't expect any of these to be very long requests. The only slightly longer(where we submit a set of translations) would be under celery tasks, so the only issue i can see happening would be an intermittent timeout which should resolve on retry otherwise it would be a bug that would need to be addressed.

esoergel commented on urls.py:
```diff
@@ -84,7 +84,8 @@
     url(r'^zipline/', include('custom.zipline.urls')),
-    url(r'^remote_link/', include('corehq.apps.linked_domain.urls'))
+    url(r'^remote_link/', include('corehq.apps.linked_domain.urls')),
+    url(r'^translations/', include('corehq.apps.translations.urls')),
```
Was this committed by mistake?  (asking because it doesn't fit with the rest of the commit)  If it's intentional, I think the translations urls are referenced in two places.

mkangia replied:
Was by mistake in an earlier commit which was then removed but now its expected change.

mkangia commented:
@esoergel bad commits, fixing. thanks for flagging

mkangia committed:
use prefetched resources

mkangia committed:
add empty line after summary for documentation

mkangia committed:
generate translation files as a part of init itself

millerdev commented on corehq/apps/app_manager/app_translations/generators.py:
```diff
@@ -190,21 +192,30 @@ def _metadata(self):
 
     def generate_translation_files(self):
         self._build_translations()
```
Just pointing out that this object has the same pattern that `PoFileGenerator` had before acdf4f4812b58c9196f9ffe7c091af4cff145dcb: it's in a partially initialized state until `generate_translation_files()` is called. I'm not going to ask you to re-architect it in this PR, but it's some food for thought.

mkangia committed:
introduce TranslationsTab to manage sidebar and better naming

esoergel commented on corehq/apps/translations/urls.py:
```diff
@@ -7,9 +7,9 @@
+        name=ConvertTranslations.urlname),
     url(r'^pull_resource/$', PullResource.as_view(),
         name=PullResource.urlname),
-    url(r'^$', ConvertTranslations.as_view(),
```
Note: you can also use something like this if you want a default for `r'^$'`:
```python
url(r'^$', RedirectView.as_view(url=ConvertTranslations.urlname, permanent=True), name='translations_main'),
```

esoergel commented on corehq/tabs/tabclasses.py:
```diff
@@ -1420,6 +1427,35 @@ def sidebar_items(self):
         return items
 
 
+class TranslationsTab(UITab):
```
:+1:

esoergel commented on custom/icds/translations/integrations/transifex.py:
```diff
@@ -39,22 +39,20 @@ def __init__(self, domain, app_id, source_lang, project_slug, version=None, lang
-            t, v, tb = sys.exc_info()
-            self._cleanup()
-            six.reraise(t, v, tb)
+        return file_uploads
```
**[non-blocking]**
I think you commented on this somewhere, but this is another example of what @millerdev was highlighting.  `TransifexPOFileGenerator` has 4 interactions with this client code - initialization, generation, file access, and cleanup.  What makes this confusing is that they are all order-dependent.  You'd get an error if you attempted to do any of that out of order, and it wouldn't make sense to do any less than all 4 operations.  For that reason, my preference would be to write the interface accordingly.

This means one step which accepts all necessary parameters, generates the files, and returns the result.  That could be done with a simple function, but it still leaves cleanup.  Cleanup is something that needs to always happen, and you can trust the calling code to handle it in a `finally` block like you've done here, but python's context managers are meant specifically for this pattern.  The way I think of it, is that you're allocating resources, which should be signified with a `with` statement that'll handle allocation and deallocation cleanly and automatically.

```python
with TransifexPoFileGenerator(...) as po_files:
    # do something with po_files
```

If implementing that is more trouble than it's worth, you could also opt to constrain `transifex_po_file_generator` to just one method so you can see it's whole lifecycle:

```python
def send_translation_files(self):
    try:
        transifex_po_file_generator = TransifexPoFileGenerator(...)
        transifex_po_file_generator.generate_translation_files()
        file_uploads = self._send_files_to_transifex(
            transifex_po_file_generator.generated_files)
    finally:
        transifex_po_file_generator._cleanup()
    return file_uploads
```
This way nothing outside of that method needs to know about `transifex_po_file_generator`, and you don't have to worry about multiple methods accessing a mutating shared state.

That's all the pontificating I'm gonna do on the topic, thank you for your patience.  I'm not expecting you to act on these suggestions, I'm more thinking of this as food for thought for next time, and options if you end up deciding to refactor some of this anyways.  Hope it's helpful.

mkangia committed:
add ability to update resource

add ability for the user to define when they want to update resource and not upload it.
